File: src/fintendo/models/research.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field
from fintendo.models.technical_confidence import TechnicalConfidence
from fintendo.models.technical_evidence import TechnicalEvidence
from fintendo.models.market_intelligence import MarketIntelligenceAnalysis



class Evidence(BaseModel):
    source: str = Field(min_length=1)
    title: str = Field(min_length=1)
    content: str = Field(min_length=1)
    timestamp: datetime

class FundamentalEvidence(BaseModel):
    source: str = Field(min_length=1)
    field: str = Field(min_length=1)
    value: float | None = None
    period: datetime
    description: str = Field(min_length=1)

class FundamentalAnalysis(BaseModel):
    ticker: str = Field(min_length=1)
    score: float = Field(ge=0, le=100)
    summary: str = Field(min_length=1)
    strengths: list[str]
    weaknesses: list[str]
    catalysts: list[str]
    risks: list[str]
    evidence: list[FundamentalEvidence]
    confidence: float = Field(ge=0, le=1)


class TechnicalAnalysis(BaseModel):
    ticker: str = Field(min_length=1)
    score: float = Field(ge=0, le=100)
    summary: str = Field(min_length=1)
    trend: Literal["bullish", "bearish", "neutral", "mixed"]
    momentum: Literal["bullish", "bearish", "neutral", "mixed"]
    volatility: Literal["low", "moderate", "high"]
    bullish_signals: list[str]
    bearish_signals: list[str]
    support_levels: list[float]
    resistance_levels: list[float]
    outlook: str = Field(min_length=1)
    evidence: list[TechnicalEvidence]
    confidence: TechnicalConfidence

class SentimentAnalysis(BaseModel):
    ticker: str = Field(min_length=1)
    score: float = Field(ge=0, le=100)
    summary: str = Field(min_length=1)
    sentiment: str = Field(min_length=1)
    key_themes: list[str]
    positive_signals: list[str]
    negative_signals: list[str]
    risks: list[str]
    evidence: list[Evidence]
    confidence: float = Field(ge=0, le=1)

class CommitteeDecision(BaseModel):
    ticker: str = Field(min_length=1)
    recommendation: str = Field(min_length=1)
    conviction: float = Field(ge=0, le=100)
    rationale: str = Field(min_length=1)
    bull_case: list[str]
    bear_case: list[str]
    key_risks: list[str]
    confidence: float = Field(ge=0, le=1)

class ResearchReport(BaseModel):
    ticker: str = Field(min_length=1)
    generated_at: datetime
    fundamental: FundamentalAnalysis
    technical: TechnicalAnalysis
    market_intelligence: MarketIntelligenceAnalysis | None
    committee: CommitteeDecision

class FundamentalAnalysisOutput(BaseModel):
    ticker: str = Field(min_length=1)
    score: float = Field(ge=0, le=100)
    summary: str = Field(min_length=1)
    strengths: list[str]
    weaknesses: list[str]
    catalysts: list[str]
    risks: list[str]
    

File: frontend/types/research.ts
export type TechnicalEvidenceType = "observed" | "derived" | "level";

export interface TechnicalConfidence {
  score: number;
  observations: number;
  history_score: number;
  trend_score: number;
  momentum_score: number;
  volatility_score: number;
  levels_score: number;
  missing_indicators: string[];
  limitations: string[];
}

export interface TechnicalEvidence {
  source: string;
  field: string;
  evidence_type: TechnicalEvidenceType;
  value: number;
  period: string;
  description: string;
}

export interface FundamentalEvidence {
  source: string;
  field: string;
  value: number | null;
  period: string;
  description: string;
}

export interface FundamentalAnalysis {
  ticker: string;
  score: number;
  summary: string;
  strengths: string[];
  weaknesses: string[];
  catalysts: string[];
  risks: string[];
  evidence: FundamentalEvidence[];
  confidence: number;
}

export type TechnicalTrend = "bullish" | "bearish" | "neutral" | "mixed";
export type TechnicalMomentum = "bullish" | "bearish" | "neutral" | "mixed";
export type TechnicalVolatility = "low" | "moderate" | "high";

export interface TechnicalAnalysis {
  ticker: string;
  score: number;
  summary: string;
  trend: TechnicalTrend;
  momentum: TechnicalMomentum;
  volatility: TechnicalVolatility;
  bullish_signals: string[];
  bearish_signals: string[];
  support_levels: number[];
  resistance_levels: number[];
  outlook: string;
  evidence: TechnicalEvidence[];
  confidence: TechnicalConfidence;
}

export type Sentiment =
  | "positive"
  | "negative"
  | "neutral"
  | "mixed";

export type SentimentStrength = "weak" | "moderate" | "strong";

export type Materiality = "low" | "moderate" | "high";

export type ImpactDirection =
  | "positive"
  | "negative"
  | "neutral"
  | "uncertain";

export type TimeHorizon =
  | "short_term"
  | "medium_term"
  | "long_term"
  | "uncertain";

export interface NewsArticle {
  source: string;
  title: string;
  url: string;
  published_at: string | null;
  author: string | null;
  content: string;
  ticker: string;
}

export interface MarketEvent {
  ticker: string;
  event_type: string;
  title: string;
  summary: string;
  sentiment: Sentiment;
  sentiment_strength: SentimentStrength;
  materiality: Materiality;
  potential_impact: ImpactDirection;
  affected_areas: string[];
  time_horizon: TimeHorizon;
  entity_scope: string;
  source: string;
  source_url: string;
  published_at: string | null;
}

export interface MarketIntelligenceAnalysis {
  ticker: string;
  score: number;
  overall_sentiment: Sentiment;
  sentiment_strength: SentimentStrength;
  summary: string;
  key_events: MarketEvent[];
  positive_factors: string[];
  negative_factors: string[];
  catalysts: string[];
  risks: string[];
  short_term_outlook: ImpactDirection;
  medium_term_outlook: ImpactDirection;
  long_term_outlook: ImpactDirection;
}

export interface Evidence {
  source: string;
  title: string;
  content: string;
  timestamp: string;
}

export interface SentimentAnalysis {
  ticker: string;
  score: number;
  summary: string;
  sentiment: string;
  key_themes: string[];
  positive_signals: string[];
  negative_signals: string[];
  risks: string[];
  evidence: Evidence[];
  confidence: number;
}

export interface CommitteeDecision {
  ticker: string;
  recommendation: string;
  conviction: number;
  rationale: string;
  bull_case: string[];
  bear_case: string[];
  key_risks: string[];
  confidence: number;
}

export interface ResearchReport {
  ticker: string;
  generated_at: string;
  fundamental: FundamentalAnalysis;
  technical: TechnicalAnalysis;
  market_intelligence: MarketIntelligenceAnalysis | null;
  committee: CommitteeDecision;
}